src/main/java/tech/indicio/ariesmobileagentandroid/messaging/transports/WSClientTransport.java

```diff
@@ -0,0 +1,132 @@
+        private final HashMap<String, ConnectionRecord> connections;
+        private final boolean failed; //If the listener has previously failed
+        private boolean hasConnected = false;
+
+        private AriesSocketListener(String endpoint, HashMap<String, ConnectionRecord> connections, boolean failed) {
+            this.endpoint = endpoint;
+            this.connections = connections;
+            this.failed = failed;
+        }
+
+        public AriesSocketListener(String endpoint, HashMap<String, ConnectionRecord> connections) {
+            this(endpoint, connections, false);
+        }
+
+        public void onClosed(@NotNull WebSocket webSocket, int code, @NotNull String reason) {
+            Log.d(TAG, "Socket for endpoint " + endpoint + " has been closed: " + code + ": " + reason);
+        }
+
+        public void onClosing(@NotNull WebSocket webSocket, int code, @NotNull String reason) {
+            super.onClosing(webSocket, code, reason);
+        }
+
+
+        public void onFailure(@NotNull WebSocket webSocket, @NotNull Throwable t, @Nullable Response response) {
+            super.onFailure(webSocket,t,response);
+            if(hasConnected){
+                Log.d(TAG, "Websocket failure for endpoint: " + endpoint + "\n" + t.getMessage());
+
+                //OnFail recreate socket
+                socketMap.remove(endpoint);
+                OkHttpClient client = new OkHttpClient();
+                Request request = new Request.Builder().url(endpoint).build();
+                AriesSocketListener listener = new AriesSocketListener(endpoint, connections, true);
+                WebSocket ws = client.newWebSocket(request, listener);
+                socketMap.put(endpoint, new Pair<>(ws, listener));
+            }
+        }
+
+        public void onMessage(@NotNull WebSocket webSocket, @NotNull String message) {
+            Log.d(TAG, "WS String message received for endpoint: "+endpoint);
+            messageReceiver.receiveMessage(message.getBytes());
+        }
+
+        public void onMessage(@NotNull WebSocket webSocket, @NotNull ByteString message) {
+            Log.d(TAG, "WS ByteString message received for endpoint: "+endpoint);
+            messageReceiver.receiveMessage(message.toByteArray());
+        }
+
+        public void onOpen(@NotNull WebSocket webSocket, @NotNull Response response) {
+            this.hasConnected = true;
+            Log.d(TAG, "Socket opened for endpoint: " + endpoint);
+            if (this.failed) {
+                Log.d(TAG, "Sending trust pings");
+                TrustPingMessage tp = new TrustPingMessage("all");
+                for (ConnectionRecord connection : this.connections.values()) {
+                    try {
+                        messageSender.sendMessage(tp, connection);
+                    } catch (Exception e) {
+                        e.printStackTrace();
+                    }
+                }
+                Log.d(TAG, "Trust pings sent");
```

Comment:
I would quite like to move the trust ping creation outside of the transport, but that can be a later effort. Would you please add a TODO here for our/other's reference?